exposed-core/src/main/kotlin/org/jetbrains/exposed/sql/statements/BatchUpsertStatement.kt

```diff
@@ -0,0 +1,39 @@
+package org.jetbrains.exposed.sql.statements
+
+import org.jetbrains.exposed.sql.Column
+import org.jetbrains.exposed.sql.Expression
+import org.jetbrains.exposed.sql.Table
+import org.jetbrains.exposed.sql.Transaction
+import org.jetbrains.exposed.sql.vendors.H2Dialect
+import org.jetbrains.exposed.sql.vendors.H2FunctionProvider
+import org.jetbrains.exposed.sql.vendors.MysqlFunctionProvider
+
+/**
+ * Represents the SQL command that either batch inserts new rows into a table, or updates the existing rows if insertions violate unique constraints.
+ *
+ * @param table Table to either insert values into or update values from.
+ * @param keys (optional) Columns to include in the condition that determines a unique constraint match. If no columns are provided,
+ * primary keys will be used. If the table does not have any primary keys, the first unique index will be attempted.
+ * @param onUpdate List of pairs of specific columns to update and the expressions to update them with.
+ * If left null, all columns will be updated with the values provided for the insert.
+ * @param shouldReturnGeneratedValues Specifies whether newly generated values (for example, auto-incremented IDs) should be returned.
+ * See [Batch Insert](https://github.com/JetBrains/Exposed/wiki/DSL#batch-insert) for more details.
+ */
+open class BatchUpsertStatement(
+    table: Table,
+    vararg val keys: Column<*>,
+    val onUpdate: List<Pair<Column<*>, Expression<*>>>?,
+    shouldReturnGeneratedValues: Boolean = true
+) : BaseBatchInsertStatement(table, ignore = false, shouldReturnGeneratedValues) {
+
```

Comment:
**Note**: Single statement [`upsert()`](https://github.com/JetBrains/Exposed/blob/master/exposed-core/src/main/kotlin/org/jetbrains/exposed/sql/statements/UpsertStatement.kt#L20) takes a WHERE clause as an argument (to be used in update part), but this option hasn't been included in `batchUpsert()`.

The rationale is that I couldn't find a use case for a batch upsert including a where that applies to all batch data (it seems to defeat the purpose?). All the PostgreSQL examples I came across used an UPDATE without WHERE (and MySQL doesn't allow WHERE for even single upsert).

I can implement its inclusion if anybody believes it should definitely still be an option. Or we can wait to see if users request it.

Reply:
We can note this in KDoc with link to youtrack if there is a use-case